quantized_mesh_encoder/extensions.py

```diff
@@ -0,0 +1,46 @@
+import abc
+from dataclasses import dataclass, field
+from enum import IntEnum
+from struct import pack
+
+import numpy as np
+
+from .constants import EXTENSION_HEADER, WGS84
+from .ellipsoid import Ellipsoid
+from .normals import compute_vertex_normals, oct_encode
+
+
+class ExtensionId(IntEnum):
+    VERTEX_NORMALS = 1
+    WATER_MASK = 2
+    METADATA = 4
+
+
+@dataclass
+class ExtensionBase(metaclass=abc.ABCMeta):
+    id: ExtensionId = field(init=False)
+
+    @property
+    @abc.abstractmethod
+    def encode(self) -> bytes:
+        """"Return the encoded extension data"""
+        ...
+
+
+@dataclass
+class VertexNormalsExtension(ExtensionBase):
+    id: ExtensionId = field(init=False, default=ExtensionId.VERTEX_NORMALS)
```

Comment:
Why do you need the `init=False` here or above? I thought `init=False` was to declare that something is ok to not be initialized at the start. So in the base I think you can just have
```py
class ExtensionBase(metaclass=abc.ABCMeta):
    id: ExtensionId
```

and here just have
```py
class VertexNormalsExtension(ExtensionBase):
    id: ExtensionId = ExtensionId.VERTEX_NORMALS
```
?

Reply:
Originally I had the following:
```python
@dataclass
class ExtensionBase(metaclass=abc.ABCMeta):
    id: ExtensionId

    @property
    @abc.abstractmethod
    def encode(self) -> bytes:
        """"Return the encoded extension data"""
        ...


@dataclass
class VertexNormalsExtension(ExtensionBase):
    id: ExtensionId = ExtensionId.VERTEX_NORMALS
    indices: np.ndarray
    positions: np.ndarray
    ellipsoid: Ellipsoid = WGS84
```
but that was causing the following error:

```
TypeError: non-default argument 'indices' follows default argument
```

Setting `init=False` made the error go away and also seemed reasonable since somebody isn't going to be passing in their own id value when calling `VertexNormalsExtension()`

Reply:
> ```
> TypeError: non-default argument 'indices' follows default argument
> ```

That just means you need to reorder them. It's just like a function where you can't have

```py
def fn(a=1, b):
	pass
#   File "<ipython-input-97-8aa6484e15de>", line 1
#     def fn(a=1, b):
#            ^
# SyntaxError: non-default argument follows default argument
```

Switch to the following and the error will go away

```py
@dataclass
class VertexNormalsExtension(ExtensionBase):
    indices: np.ndarray
    positions: np.ndarray
    id: ExtensionId = ExtensionId.VERTEX_NORMALS
    ellipsoid: Ellipsoid = WGS84
```